### ISSUE TITLE
Combine is not creating heaps in mode 6

### ISSUE THREAD

**Issue by Datalund** [Courseplay/courseplay#1971]
Farming simulator version: 1.4.4.0
Courseplay version: V5.01.00116.dev
Map: Soznovka
Mods: Courseplay

Log file: https://gist.github.com/Datalund/2067d41aa7f353036cca25ed800ac3aa#file-gistfile1-txt

I've been trying with every possible combination of this, but no matter what I do, the combine is not creating any heap at all.
It doesn't matter if I use a combine harvester or a rootcrop harvester. The machinery will go to the Tip point and just stop. I have tried sitting up a trailer underneath the pipe, and then the combine unloads, but it wont create any heap.
As you can see it creates a huge log file, but that is just about it.
In the picture I have setup a field course for the harvester on field 19. After that course I have setup a unload course.
The two courses are setup like this: Field course loaded first and the unload course is added with the append function.

(The course in the picture is just for showing. No matter how I set it up, it won't work)

![20170411124603_1](https://cloud.githubusercontent.com/assets/27280767/24906988/85df32b0-1eba-11e7-87a2-72b73a2ecc9f.jpg)

**Comment by mixmax1988** [comment]
Correct me if i am wrong but i think in selfunload harvester mode you have to put 2waitpoints to unload in a heap. The heap will be created between the 2WP

**Comment by Datalund** [comment]
With two Wait points the combine will stop at the first and say "has reached overload point" like seen on the picture below:

![20170412014955_1](https://cloud.githubusercontent.com/assets/27280767/24935443/6b2218d0-1f22-11e7-9758-88c134db9767.jpg)

With two Unload points the combine wont even stop. It will slow down and drive on in a straight line (it wont follow the recorded course) as show on the picture below:

![20170412015333_1](https://cloud.githubusercontent.com/assets/27280767/24935545/032c6662-1f23-11e7-88f4-677f2ce1c9f8.jpg)


And according to the change log shown below, it should be done with just one Unloadpoint :)

![udklip](https://cloud.githubusercontent.com/assets/27280767/24935657/a64d8402-1f23-11e7-9a43-eca7dc858af7.JPG)

**Comment by mixmax1988** [comment]
You are right, didn't see that.. ignore my previous comment then

**Comment by Raints** [comment]
Not sure but I did try this also and it only worked when combine was hired not using CP for it.

**Comment by Tensuko** [comment]
In 'self-unload' mode, the only Course you create is the way from the field to the unload position. You hire a normal helper for Harvesting and then hire the CP helper. Then it should work fine.

**Comment by Datalund** [comment]
You are right. It works if I use a hired helper and run the "unload course" in mode 7, but then what is the point of line 7 and 8 in the changelog 01.00101? Why write that we can now create heaps in mode 6 for combines and rootcrop harvesters if we still need to use mode 7 for it? And furthermore in mode 7 it still requires you to set two unloadpoints for it to work. Otherwise you'll get the "Too few unload points! 2 needed" error as shown in the picture below.

![20170412125937_1](https://cloud.githubusercontent.com/assets/27280767/24954573/38e1b688-1f80-11e7-833e-6ff84ab008dd.jpg)


If you try to run the "unload course" in mode 6 where the fieldwork is done by a hired helper, you end up with a combine that is sliding when it has to turn at the end of the field and it won't work.

So even though we can use the hired helper for the fieldwork, and setup a course for unloading in mode 7 with two unloadpoints, I still believe there is a bug :) 
Besides the ingame hired helper kinda requires us to have perfectly square fields for it to work as well

**Comment by Satissis** [comment]
> furthermore in mode 7 it still requires you to set two unloadpoints for it to work. Otherwise you'll get the "Too few unload points! 2 needed" error.

I know it can be a bit confusing, but here is how things works.
The difference between 1 wait point and 1 unload point is:
Wait Points: Unload in a trailer if there is one and if not, then wait until one is there.
Unload Point: Unload in a trailer if there is one and if not, then unload to ground.

I didn't think you would need to use the 1 unload point to unload to ground in mode 7, which is why it's giving that error.

To sum up on the above, then:
2 Unload points = heap unloading.
1 Wait point = unload in a trailer if there is one and if not, then wait until one is there.
1 Unload point, do not work in mode 7

EDIT: Doh just realized that I actually posted that unloading to ground in mode 7 was also possible.
I'll add that in as well.

**Comment by Datalund** [comment]
@Satissis Thanks for clearing that up :)

**Comment by pops64** [comment]
Try again with 145 or higher and report back. Mode 7 single point heap making is on hold or may not even be possible to fix. Mode 6 works fine in single point and double from my test

**Comment by Datalund** [comment]
Thanks for the update

I did the test in mode 6 with a combine harvester and two unload points and it passed with no problems.
The same test in mode 6 with a rootcrop harvester and two unload points failed however. The tractor stops at the first unload point and doesn't continue the course.

I recorded it all and you can see it here (the rootcrop test starts at [2:25](https://youtu.be/zLGQPcSJoFs?t=145) minuttes
https://youtu.be/zLGQPcSJoFs

**Comment by pops64** [comment]
Yea. I set it so if it is going to break to just stop instead of just slowly wandering in a circle. This is the same problem I have with mode 7. The giants function to check to see if discharge is possible has some conditions in it. That make it reutrn false even though it should be true. Do you have a log to make sure it is not a nil error.

**Comment by Datalund** [comment]
Clever idea to make it stop instead of having the machine drive around in a circle :)

I've uploaded the log to gist:
https://gist.github.com/Datalund/337bfa06877ac2050f91a0786c8d701e

**Comment by pops64** [comment]
Try again with version 164. And report back. Mode 6 know works with all combines using unload points 1 and 2. Mode 7 stays the same with only 2 points usable. For unknown reasons, mode 7 doesn't unload with 1

**Comment by Datalund** [comment]
Tested with version 165.

- Map: Goldcrest Valley
- Vanilla modfolder with only Courseplay and Follow Me
- Courses was setup for both harvesting and unloading
- Unload course was setup with 2 unload points.
Self propelled haversters (combine harvesters, potato harvesters and beet harvesters) passed with no issues at all in mode 6.

Same conditions with only one unload point in mode 6:
Self propelled haversters (combine harvesters, potato harvesters and beet harvesters) passed with no issues at all in mode 6.

Test with trailed harvesters for potato and beets in mode 6 failed however. I get the same result as shown in my video a couple of comments up (https://youtu.be/zLGQPcSJoFs)

**Comment by pops64** [comment]
Try again with version 166 or higher and report back thanks. On my todo list is to move the pipe unload point to where the unload point is for 2 unload points. 1 unload point will unload where you place the unload point

**Comment by Datalund** [comment]
Tested with version 167 and both self propelled harvesters and trailed harvesters works perfectly now! You 
are right about the unload point is set to the tractor when you place 2 unload points, but that is a minor issue and nothing that can't be worked around ;)
Unload point is placed perfectly at the pipe of the machine when you only place one unload point.
You have done an amazing job fixing this bug. Thank you so much :)

**Comment by pops64** [comment]
Your welcome :D. I am not sure if I will ever be able to center the pipe with 2 points because of the function to unload is called. I will look into later. Closed to keep issues area clean.
